### PR TITLE
chore(ui-react, ui): expose Authenticator types, add AmplifyUser interface

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes
Expose `Authenticator` types for end user consumption:
- rename `AuthenticatorHOCProps` as `WithAuthenticatorProps` and publicly export
- publicly export `AuthenticatorProps` and `WithAuthenticatorOptions`
- add `AmplifyUser` interface to replace `CognitoUserAmplify`, mark `CognitoUserAmplify` as deprecated
- remove fragile `CognitoUserAmplify` and `CodeDeliveryDetails` links from the docs

#### Issue #, if available
NA

#### Description of how you validated changes
Tested in example apps

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are updated
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/src/pages/[platform]/connected-components/authenticator/headless/headless-usage.web.mdx [hunk 1]
```diff
@@ -56,7 +56,7 @@ Here's an example that uses the `toResetPassword` trigger transition, to create
         <TableRow>
           <ResponsiveTableCell label="Name">`user`</ResponsiveTableCell>
           <ResponsiveTableCell label="Description">Current signed in user</ResponsiveTableCell>
-          <ResponsiveTableCell label="Type">[`CognitoUserAmplify`](https://github.com/aws-amplify/amplify-ui/blob/f9be291e4b89137d7b521bd3e82b4f96ee4323f6/packages/ui/src/types/authMachine.ts#L113-L115)</ResponsiveTableCell>
```
Comment (calebpollman) [inline docs/src/pages/[platform]/connected-components/authenticator/headless/headless-usage.web.mdx:59]: Removed this link as it points to a commit, in a file that no longer exists in `main`

File: docs/src/pages/[platform]/connected-components/authenticator/headless/headless-usage.web.mdx [hunk 2]
```diff
@@ -86,7 +86,7 @@ Here's an example that uses the `toResetPassword` trigger transition, to create
         <TableRow>
           <ResponsiveTableCell label="Name">`codeDeliveryDetail`</ResponsiveTableCell>
           <ResponsiveTableCell label="Description">Provides detail on where confirm sign up code is sent to.</ResponsiveTableCell>
-          <ResponsiveTableCell label="Type">[`CodeDeliveryDetail`](https://github.com/aws-amplify/amplify-js/blob/6c4d4b5fc737652f88dcdfa26bdb02a1defdfa2e/packages/amazon-cognito-identity-js/index.d.ts#L9-L13)</ResponsiveTableCell>
```
Comment (calebpollman) [inline docs/src/pages/[platform]/connected-components/authenticator/headless/headless-usage.web.mdx:89]: Removed this link as it points to the Amplify JS repo, which we do not maintain and could potentially change without us being aware

File: .changeset/thick-bears-buy.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+---
+"docs": patch
```
Comment (slaymance) [inline .changeset/thick-bears-buy.md:2]: `docs` shouldn't be included as part of this changeset as we only want to bump public packages.

**Review by slaymance — CHANGES_REQUESTED** [review]

File: .changeset/thick-bears-buy.md (hunk 3 above)
Comment (calebpollman) [reply .changeset/thick-bears-buy.md:2]: I add what the default changeset provides 😄  (we should adjust the settings)

Comment (calebpollman) [inline .changeset/thick-bears-buy.md:2]:
```suggestion
````

**Review by slaymance — APPROVED** [review]

**Review by reesscot — APPROVED** [review]
LGTM

**Review by wlee221 — APPROVED** [review]

**Pull request merged**